### PR TITLE
save dataframe and model metdata in HDFS

### PR DESCRIPTION
**Pull request by vansika**
# Problem
- dataframe metadata is not being saved
- model metdata is not being saved
- using [a json file ](https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz_spark/recommendations/recommendation-metadata.json) to track model id of model saved in HDFS.

# Solution

- save dataframe metadata
- save model metadata
- track model id from metadata parquets instead of using a json file.

# Action

dataframe metadata schema
- dataframe_created
- dataframe_id
- from_date
- listens_count
- playcounts_count
- recordings_count 
- to_date
- users_count

model metadata will always store the most recently created dataframe id because this dataframe will be used to train the model

model metdata schema
- dataframe_id
- model_created
- model_id  
- alpha
- lmbda
- iteration
- rank
- test_data_count
- test_rmse
- training_data_count
 - validation_data_count
 - validation_rmse

recommend.py will use the most recently saved model id to load the model.
For now, at any point in time, only one model will be saved in HDFS and the previous model will be deleted.

### CODE REVIEW

**Comment by vansika** [comment]
@brainzbot retest this please

**Comment by vansika** [comment]
@mayhem it is a big PR but everything is interlinked so it was best to make all the changes in one PR. The description should be good enough to understand the changes.

**Comment by mayhem** [comment]
Can you please resolve the merge conflicts first?

**Comment by vansika** [comment]
Whoops. Missed that. Sure.

File: listenbrainz_spark/recommendations/tests/test_models.py [hunk 1]
```diff
@@ -22,44 +25,158 @@ def tearDownClass(cls):
...
-        self.assertEqual(rating_object.product, 1)
-        self.assertEqual(rating_object.rating, 1)
+        self.assertEqual(rating_object.product, 2)
+        self.assertEqual(rating_object.rating, 3)
+
+    @patch('listenbrainz_spark.recommendations.train_models.sqrt')
+    @patch('listenbrainz_spark.recommendations.train_models.RDD')
+    @patch('listenbrainz_spark.recommendations.train_models.train')
+    @patch('listenbrainz_spark.recommendations.train_models.add')
+    def test_compute_rmse(self, mock_add, mock_train, mock_rdd, mock_sqrt):
```
Comment (vansika) [inline listenbrainz_spark/recommendations/tests/test_models.py:39]: @mayhem  I would like a review on this test especially.

File: listenbrainz_spark/recommendations/tests/test_models.py [hunk 2]
```diff
@@ -22,44 +25,158 @@ def tearDownClass(cls):
...
+        n = 1
+        model_id = "281c4177-f33a-441d-b15d-910acaf18b07"
+        _ = train_models.compute_rmse(mock_train.return_value, mock_rdd, n, model_id)
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/tests/test_models.py:42]: I dont understand this -- you are using the return_value of the mock, but the mock hasn't been used yet...

File: listenbrainz_spark/recommendations/tests/test_models.py [hunk 3]
```diff
@@ -22,44 +25,158 @@ def tearDownClass(cls):
...
+
+        mock_predict_all = mock_train.return_value.predictAll
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/tests/test_models.py:44]: and what does this do?

File: listenbrainz_spark/recommendations/tests/test_models.py [hunk 4]
```diff
@@ -22,44 +25,158 @@ def tearDownClass(cls):
...
+            rmse_time="2.1",
+            alpha=3.0,
         )
-        model_save_path = os.path.join('/test/model', best_model_metadata['model_id'])
-        train_models.save_model(model_save_path, best_model_metadata['model_id'], best_model)
-        model_exist = utils.path_exists(model_save_path)
-        self.assertTrue(model_exist)
+
+        metadata = train_models.get_best_model_metadata(best_model)
+        self.assertEqual(best_model.validation_rmse, metadata['validation_rmse'])
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/tests/test_models.py:110]: is metadata['validation_rmse'] the same as the constant value of 3.9 from above?

File: listenbrainz_spark/recommendations/train_models.py [hunk 5]
```diff
@@ -69,8 +80,88 @@ def preprocess_data(playcounts_df):
...
 
-def train(training_data, validation_data, num_validation, ranks, lambdas, iterations, alpha):
-    """ Train the data and get models as per given parameters i.e. ranks, lambdas and iterations.
+def generate_model_id():
+    """ Generate model id.
+    """
+    return '{}-{}'.format(config.MODEL_ID_PREFIX, uuid.uuid4())
+
+
+def get_best_model_path(model_id):
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/train_models.py:89]: I can confused by this name. why best when there are not evaluations of quality being made in this function?

File: listenbrainz_spark/recommendations/train_models.py [hunk 6]
```diff
@@ -69,8 +80,88 @@ def preprocess_data(playcounts_df):
...
+            model_id (str): Model identification string of best model.
+
+        Returns:
+            path to save best model.
+    """
+
+    return config.HDFS_CLUSTER_URI + path.DATA_DIR + '/' + model_id
+
+
+def get_latest_dataframe_id(dataframe_metadata_df, best_model_metadata):
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/train_models.py:102]: This function is named get_ ... but it doesn't return anything. Nor does the docstring say that data in best_model_metadata will be updated.

File: listenbrainz_spark/recommendations/recommend.py [hunk 7]
```diff
@@ -9,22 +9,51 @@
...
 from flask import current_app
+import pyspark.sql.functions as func
 from pyspark.sql.functions import col
 from pyspark.sql.utils import AnalysisException
 from pyspark.mllib.recommendation import MatrixFactorizationModel
 
 
-def load_model(path):
-    """ Load best model from given path in HDFS.
+def get_best_model_id():
```
[the remark this answers is not in the document]
Comment (vansika) [reply listenbrainz_spark/recommendations/recommend.py:23]: we have been using best model everywhere. It is like the model chosen from a group of models which will be eventually saved and loaded. 
umm....so what do you think? should we go for get_most_recent_model_id?

File: listenbrainz_spark/recommendations/tests/test_models.py (hunk 2 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/tests/test_models.py:42]: an actual model should be passed to `compute_rmse`.
the model is the return value of `train_model` function and not some kind of import that I can directly patch.

is there any other way to do that?

File: listenbrainz_spark/recommendations/tests/test_models.py (hunk 3 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/tests/test_models.py:44]: This test overall ensures that all the methods in [compute_rmse](https://github.com/metabrainz/listenbrainz-server/blob/master/listenbrainz_spark/recommendations/train_models.py#L39) are called in order wherein we are passing return value of a method to the next method.

specifically here, `predictAll` is being called on the return value of `train_model` which is an actual model.

File: listenbrainz_spark/recommendations/tests/test_models.py (hunk 4 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/tests/test_models.py:110]: yes. we are converting nametuple to a dict in `get_best_model_metadata`

File: listenbrainz_spark/recommendations/train_models.py (hunk 5 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/train_models.py:89]: right. so initially we are training the model and getting the best one out of it. Later everywhere we are referring to that model. but I guess it's confusing.
So we should just say model path, right?

File: listenbrainz_spark/recommendations/train_models.py (hunk 6 above)
Comment (vansika) [reply listenbrainz_spark/recommendations/train_models.py:102]: oops! should be `update_best_model_metadata`?

Comment (mayhem) [reply listenbrainz_spark/recommendations/train_models.py:102]: Sounds better, yes.

File: listenbrainz_spark/recommendations/train_models.py (hunk 5 above)
Comment (mayhem) [reply listenbrainz_spark/recommendations/train_models.py:89]: Yes.

File: listenbrainz_spark/recommendations/tests/test_models.py (hunk 2 above)
Comment (mayhem) [reply listenbrainz_spark/recommendations/tests/test_models.py:42]: Not really sure. I think having @paramsingh take a look would be good.

Comment (paramsingh) [reply listenbrainz_spark/recommendations/tests/test_models.py:42]: Passing a `MagicMock` object would be clearer.

```
mock_model = MagicMock()
_ = train_models.compute_rmse(mock_model, mock_rdd, n, model_id)
```

Comment (vansika) [reply listenbrainz_spark/recommendations/tests/test_models.py:42]: Thanks :)

File: listenbrainz_spark/recommendations/train_models.py [hunk 8]
```diff
@@ -84,41 +175,95 @@ def train(training_data, validation_data, num_validation, ranks, lambdas, iterat
...
+
+
+def save_best_model(model_id, model):
+    """ Save best model to HDFS.
+
+        Args:
+            model_id (str): Best model identification string.
+            model: Trained best model
+    """
+    delete_best_model()
```
Comment (mayhem) [inline listenbrainz_spark/recommendations/train_models.py:228]: Why is this here? This seems out of place for this function...

Comment (vansika) [reply listenbrainz_spark/recommendations/train_models.py:228]: I forgot add a comment. Previous model should be deleted before saving the new one.

**Comment by vansika** [comment]
@paramsingh ready for merge
